src/runtime/virtcontainers/qemu.go

```diff
@@ -1020,47 +1049,138 @@ func (q *qemu) togglePauseSandbox(pause bool) error {
 }
 
 func (q *qemu) qmpSetup() error {
-	q.qmpMonitorCh.Lock()
-	defer q.qmpMonitorCh.Unlock()
+	return q.startQMPInstance(&q.qmpMonitorCh, "monitor-qmp", true)
+}
 
-	if q.qmpMonitorCh.qmp != nil {
+func (q *qemu) startQMPInstance(qmpChan *qmpChannel, subsystem string, enableEvents bool) error {
+	qmpChan.Lock()
+	defer qmpChan.Unlock()
+
+	if qmpChan.qmp != nil {
 		return nil
 	}
 
-	cfg := govmmQemu.QMPConfig{Logger: newQMPLogger()}
+	var events chan govmmQemu.QMPEvent
+
+	if enableEvents {
+		events = make(chan govmmQemu.QMPEvent)
+		go q.loopQMPEvent(events)
+	}
+
+	// It's safe to pass a nil EventCh
+	cfg := govmmQemu.QMPConfig{
+		Logger: qmpLogger{
+			logger: virtLog.WithField("subsystem", subsystem),
+		},
+		EventCh: events,
+	}
 
 	// Auto-closed by QMPStart().
 	disconnectCh := make(chan struct{})
 
-	qmp, _, err := govmmQemu.QMPStart(q.qmpMonitorCh.ctx, q.qmpMonitorCh.path, cfg, disconnectCh)
+	qmp, _, err := govmmQemu.QMPStart(qmpChan.ctx, qmpChan.path, cfg, disconnectCh)
 	if err != nil {
 		q.Logger().WithError(err).Error("Failed to connect to QEMU instance")
 		return err
 	}
 
-	err = qmp.ExecuteQMPCapabilities(q.qmpMonitorCh.ctx)
+	err = qmp.ExecuteQMPCapabilities(qmpChan.ctx)
 	if err != nil {
 		qmp.Shutdown()
 		q.Logger().WithError(err).Error(qmpCapErrMsg)
 		return err
 	}
-	q.qmpMonitorCh.qmp = qmp
-	q.qmpMonitorCh.disconn = disconnectCh
+	qmpChan.qmp = qmp
+	qmpChan.disconn = disconnectCh
+
+	return nil
+}
+
+func (q *qemu) loopQMPEvent(event chan govmmQemu.QMPEvent) {
+	for {
+		select {
+		case e, open := <-event:
+			if !open {
+				q.Logger().Infof("QMP event channel closed")
+				return
+			}
+			q.Logger().Debugf("got new QMP event %+v", e)
+			if e.Name == "GUEST_PANICKED" {
+				go q.handleGuestPanic()
+			}
+		}
+	}
+}
+
+func (q *qemu) handleGuestPanic() {
+	if q.config.GuestMemoryDumpEnabled {
+		q.dumpGuestMemory()
+	}
+	// TODO: how to notify the upper level sandbox to handle the error
+	// to do a fast fail(shutdown or others).
+}
+
+func (q *qemu) dumpGuestMemory() error {
+	q.Logger().Info("try to dump guest memory")
+	dumpSavePath := q.config.GuestMemoryDumpPath
+	if dumpSavePath == "" {
+		dumpSavePath = dumpSavePath
+	}
 
+	if fi, err := os.Stat(dumpSavePath); err != nil {
+		if os.IsNotExist(err) {
+			os.MkdirAll(dumpSavePath, 0755)
+		} else {
+			return err
+		}
+	} else if !fi.IsDir() {
+		return fmt.Errorf("Not a directory: %s", dumpSavePath)
+	}
+
+	protocol := fmt.Sprintf("file:%s/vmcore-%s.%s", dumpSavePath, q.id, q.config.GuestMemoryDumpFormat)
+	q.Logger().Info("try to dump guest memory to %s", protocol)
+
+	err := q.manageQmpSetup()
+	if err != nil {
+		q.Logger().WithError(err).Error("setup manage QMP failed")
+		return err
+	}
+	defer q.manageQmpShutdown()
+
+	if err := q.qmpManageCh.qmp.ExecuteDumpGuestMemory(q.qmpManageCh.ctx, protocol, q.config.GuestMemoryDumpPaging, q.config.GuestMemoryDumpFormat); err != nil {
```

Comment:
A few questions...

What happens:

- on `ENOSPC` (before the dump or when the dump is part way through)?
- if a file with the same name as the dump file already exists?
- if the dump file path start with `../`?
- if the dump dir is a sym-link?
- if `dumpGuestMemory()` gets called if that method is already running? Is this possible? Do we need any locking or does Qemu handle that?

Reply:
To be clear about these questions, the "answer" for some might turn into comments (in the code and `configuration.toml`), others might require extra code checks (?) and others might require us to create some new (security?) tests.

One other consideration: a variant of the `ENOSPC` scenarios above: we need to be careful to handle a potential DoS scenario: If the kernel panics, the chances are it will panic the next time the VM starts. So if a user or a container manager attempts to create a lot of containers very quickly, that could lead to `ENOSPC` very fast potentially. In summary: do we need to think about rate limiting here?

/cc @c3d.

Reply:
@jodh-intel  As a basic check, if available space is under double of the VM  size  (`default_memory` in `configuration.toml` for `[hypervisor]`) + `HotpluggedMemory`, then dump operation will be skipped.